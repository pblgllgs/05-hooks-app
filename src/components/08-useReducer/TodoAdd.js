import React from 'react'
import { useForm } from '../../hooks/useForms';

export const TodoAdd = ({ handleAddTodo }) => {

    //hook que maneja el formulario, reset, handle e inicializacion del input
    const [{ description }, handleInputChange, reset] = useForm({
        description: "",
    });

    const handleSubmit = (e) => {
        e.preventDefault()
        if (description.trim().length <= 2) {
            return;
        }
        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false,
        };
        handleAddTodo(newTodo);
        reset();
    }


    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />
            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    name="description"
                    className="form-control"
                    placeholder="Aprender..."
                    autoComplete="off"
                    value={description}
                    onChange={handleInputChange}
                />
                <button className="btn btn-outline-primary" type="submit">
                    Agregar
                </button>
            </form>
        </>
    )
}
