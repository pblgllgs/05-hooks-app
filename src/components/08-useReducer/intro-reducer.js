//estado inicial
const initialState = [
    {
        id: 1,
        todo: "Comprar leche",
        done: false,
    }
];

//reducer
const todoReducer = (state = initialState, action) =>{
    if(action?.type === 'agregar'){
        return [...state, action.payload]
    }
    return state;
}

//nuevo estado
const newTodo = [
    {
        id: 2,
        todo: "Comprar leche",
        done: false,
    }
];

//accion que acompa√±a al nuevo
const agregarTodoAction = {
    type:'agregar',
    payload: newTodo
}

let todos = todoReducer();
todos = todoReducer(todos, agregarTodoAction)

console.log(todos)