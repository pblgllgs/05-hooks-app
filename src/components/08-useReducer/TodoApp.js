import React, { useEffect, useReducer } from "react";
import "./styles.css";
import { TodoAdd } from "./TodoAdd";
import { TodoList } from "./TodoList";
import { todoReducer } from "./todoReducer";

const init = () => {
    //carga de los items desde el localstorage
    return JSON.parse(localStorage.getItem("todos")) || [];
};

export const TodoApp = () => {
    //reducer para cambiar el estado del los datos
    const [todos, dispatch] = useReducer(todoReducer, [], init);
    
    //gugardar en el local storge
    useEffect(() => {
        localStorage.setItem("todos", JSON.stringify(todos));
    }, [todos]);
    //eliminacion de items
    const handleDelete = (todoId) => {
        //crea la accion
        const action = {
            type: "delete",
            payload: todoId,
        };
        //ejecucion de la accion
        dispatch(action);
    };
    //tacha el item seleccionado
    const handleToggle = (todoId) => {
        //enviamos el action dentro del dispatch y cambiamos el done 
        dispatch({
            type:'toggle',
            payload: todoId
        })
    }
    const handleAddTodo= (newTodo)=> {
        //action
        dispatch({
            type: "add",
            payload: newTodo,
        });
    }

    return (
        <div>
            <h1>TodoApp ({todos.length})</h1>
            <hr />
            <div className="row">
                <div className="col-7">
                    <TodoList 
                        todos={todos} 
                        handleDelete={handleDelete} 
                        handleToggle={handleToggle} 
                    />

                    {(todos.length === 0) &&
                        <div>
                            <h4>No hay elementos</h4>
                            <div className="alert alert-info">No se encontraron Todos!!!</div>
                        </div>
                    }
                </div>
                <div className="col-5">
                    <TodoAdd 
                        handleAddTodo={handleAddTodo} 
                    />
                </div>

            </div>
        </div>
    );
};
