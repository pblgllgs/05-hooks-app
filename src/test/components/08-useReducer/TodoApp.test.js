import { shallow, mount } from "enzyme";
import { act } from "@testing-library/react";
import { TodoApp } from "../../../components/08-useReducer/TodoApp";
import { demoTodo } from "../../fixtures/demoTodo";

describe("pruebas en <TodoApp />", () => {
    const wrapper = shallow(<TodoApp />);

    Storage.prototype.setItem = jest.fn( () => {});

    test("debe de mostrarse correctamente", () => {
        //se compara un snapshot con el codigo generado
        expect(wrapper).toMatchSnapshot();
    });

    test('debe de agregar un todo', () => {
        const wrapper = mount(<TodoApp />)
        act(()=> {
            //ejecuta dos veces la funcion para agregar un todo
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodo[0])
            wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodo[1])
        })
        expect(wrapper.find('h1').text().trim()).toBe('TodoApp (2) items')
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    });

    test('debe de eleminar un todo', () => {
        //en el componente TodoAdd llama la funcion handleAddTodo y agrega un todo
        wrapper.find('TodoAdd').prop('handleAddTodo')(demoTodo[0])
        //en la lista en el todo que se agreg√≥ recien, se llama la funcion y se elimina el todo recien agregado 
        wrapper.find('TodoList').prop('handleDelete')(demoTodo[0].id)
        //se espera que en la lista no se encuentren valores
        expect(wrapper.find('h1').text().trim()).toBe('TodoApp (0) items')
    })
    
    
});
