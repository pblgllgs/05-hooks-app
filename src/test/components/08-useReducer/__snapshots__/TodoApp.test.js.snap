// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en <TodoApp /> debe de mostrarse correctamente 1`] = `
<div>
  <h1>
    TodoApp (
    0
    ) items
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
      <div>
        <h4>
          No hay elementos
        </h4>
        <div
          className="alert alert-info"
        >
          No se encontraron Todos!!!
        </div>
      </div>
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
